dhall/src/Dhall/Parser/Expression.hs

```diff
@@ -294,7 +294,7 @@ parsers embedded = Parsers {..}
     applicationExpression = do
             f <-    (Some <$ try (_Some <* nonemptyWhitespace))
                 <|> return id
-            a <- noted importExpression_
+            a <- noted importExpression_ <?> "argument to ❰Some❱"
```

Comment:
Note that `a` isn't always the argument to `Some` – the leading `Some` is optional!

Are there parser errors where this label is shown although there is no leading `Some`? If that is impossible, it would be good to say why in a comment.

Reply:
Oh yeah, that's right.  This needs to be fixed to account for whether or not the `Some` was parsed

Reply:
How to test that, what expression can result in `a` not being a `Some` argument?

Reply:
For example in `f x`, I believe `a` will represent `f`.

Does e.g. `x === ,` trigger an error involving the new label?

Reply:
Note that there's a new `haskell-syntax-tree` command that might be useful for understanding how the parser works:

```
$ echo 'f x' | dhall haskell-syntax-tree
App 
    ( Var ( V "f" 0 ) ) 
    ( Var ( V "x" 0 ) )
```

It doesn't show the `Note`s though – maybe we should add an option for that.

Reply:
`haskell-syntax-tree` looks pretty useful :) So `f x` fails early because of `Error: Unbound variable` . However `x === ,` does reproduce the issue indeed, it fails with
```
Error: Invalid input

(stdin):1:7:
  |
1 | x === ,
  |       ^^
unexpected ",<newline>"
expecting keyword, argument to ❰Some❱, or whitespace
```

Reply:
Unfortunately the code of `applicationExpression` looks like an alien incantation to me. I guess we need to use something like `<?> if f /= id then "argument to ❰Some❱" else ""` , but that doesn't seem to work. (No instance for (Eq (Expr Src a -> Expr Src a)))

Reply:
Assuming that when Some is not parsed, `f` is _id_, then that seems to work better:
```haskell
            a <- noted importExpression_ <?> case f None of
                                               None -> ""
                                               _    -> "argument to ❰Some❱"
```
Is this correct? and how else can we test the value of `f`?

Reply:
@TristanCacqueray: There is a more accurate way to detect which alternative was selected: return an additional value from each side:

```haskell
let alternative0 = do
        f <- Some <$ try (_Some <* nonemptyWhitespace)

        return (f, True)

let alternative1 = return (id, False)

(f, isSome) <- alternative0 <|> alternative1

let adapt parser = if isSome then parser <?> "argument to ❰Some❱" else parser

a <- adapt (noted importExpression)
```

However, there's actually a simpler solution: change `"argument to ❰Some❱"` to `"function argument"`

Reply:
> However, there's actually a simpler solution: change `"argument to ❰Some❱"` to `"function argument"`

Note that `a` might itself be the function when there is no `Some` to wrap it.

I like the idea of using a different value to signal whether we found a `Some` or not though. Since we're only interested whether there was a `Some` or not, a `Bool` should be sufficient on its own. No need to keep that `id` trick.

Reply:
Thank you for the suggestions and the explanations, I'm having such a hard time to understand the do notation... I used the `isSome` value to keep the original message, because it is the explanation I meant to add in the first place.

Is there a place I could add a test for that new parser message?

Reply:
Yes, this stuff isn't exactly simple. The many operators involved make it very concise but are probably rather confusing for people who aren't used to them. [This video on combinator parsing](https://www.youtube.com/watch?v=dDtZLm7HIJs) just popped up on r/haskell – maybe it's useful for understanding e.g. the do-notation here!?

A test would indeed be great but IMHO isn't strictly required! The existing parser tests are [here](https://github.com/dhall-lang/dhall-haskell/blob/a794cd88f06a151f9f4d692b6edf35259890bdca/dhall/tests/Dhall/Test/Parser.hs). We don't have any tests for the error messages yet though.

Ultimately proper golden tests would be nice for that, but setting that up is probably quite a bit of work…

Regarding the parser change, I think it could be simplified a bit – I might do that in a follow-up PR.

Reply:
That video is indeed spot-on :) I also read http://dev.stephendiehl.com/hask/#do-notation , but this is still very confusing. Getting my hands dirty with an existing code base is helping, thanks again for the pointers.

I am looking at Dhall.Test.Util to add a `codeErrors :: Text -> Text` function, but i'm currently unable to use `getErrors` with `inputExpr`. It seems like this would be a useful utility to test such PR and prevent regression.